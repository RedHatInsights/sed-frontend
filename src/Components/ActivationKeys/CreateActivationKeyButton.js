import React from 'react';
import { Button } from '@patternfly/react-core';
import { useQueryClient } from 'react-query';
import PropTypes from 'prop-types';

const CreateActivationKeyButton = (props) => {
  const { onClick } = props;
  const queryClient = useQueryClient();
  const user = queryClient.getQueryData('user');
  const isDisabled = () => {
    return !user.rbacPermissions.canWriteActivationKeys;
  };
  return (
    <Button variant="primary" onClick={onClick} isDisabled={isDisabled()}>
      Create activation key
    </Button>
  );
};

CreateActivationKeyButton.propTypes = {
  onClick: PropTypes.func.isRequired,
};

export default CreateActivationKeyButton;
