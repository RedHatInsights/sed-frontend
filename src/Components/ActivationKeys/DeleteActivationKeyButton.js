import React from 'react';
import PropTypes from 'prop-types';
import { WriteOnlyButton } from '../WriteOnlyButton';
import { TrashIcon } from '@patternfly/react-icons';

const DeleteActivationKeyButton = ({ onClick }) => {
  return (
    <WriteOnlyButton
      variant="plain"
      onClick={onClick}
      disabledTooltip="For editing access, contact your administrator."
      enabledTooltip="Delete"
    >
      <TrashIcon />
    </WriteOnlyButton>
  );
};

DeleteActivationKeyButton.propTypes = {
  onClick: PropTypes.func.isRequired,
};

export default DeleteActivationKeyButton;
