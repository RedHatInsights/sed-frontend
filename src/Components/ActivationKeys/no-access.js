import {
  Bullseye,
  EmptyState,
  EmptyStateBody,
  EmptyStateIcon,
  Title,
} from '@patternfly/react-core';
import { LockIcon } from '@patternfly/react-icons';
import React from 'react';

const NoAccessView = () => (
  <Bullseye>
    <EmptyState>
      <EmptyStateIcon icon={LockIcon} />
      <Title headingLevel="h4" size="lg">
        Activation keys can only be accessed by organization administrators.
      </Title>
      <EmptyStateBody>
        If you already know your organization ID and activation key, you can
        register systems with RHC.
      </EmptyStateBody>
    </EmptyState>
  </Bullseye>
);

export default NoAccessView;
