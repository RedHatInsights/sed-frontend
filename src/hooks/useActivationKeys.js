import { useQuery } from '@tanstack/react-query';
import useChrome from '@redhat-cloud-services/frontend-components/useChrome';

const fetchActivationKeysData = (token) => async () => {
  const response = await fetch('/api/rhsm/v2/activation_keys', {
    headers: { Authorization: `Bearer ${await token}` },
  });

  const activationKeysData = await response.json();

  return activationKeysData.body;
};

const getActivationKeys = (token) => async () => {
  const keysData = await fetchActivationKeysData(token)();
  return keysData;
};

const useActivationKeys = () => {
  const chrome = useChrome();

  return useQuery(['activation_keys'], getActivationKeys(chrome?.auth?.getToken()));
};

export { useActivationKeys as default };
