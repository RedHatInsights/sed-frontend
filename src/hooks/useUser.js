import { useQuery } from 'react-query';
import { useRbacPermissions } from './useRbacPermissions';
import { useAuthenticateUser } from '../utils/platformServices';

const useUser = () => {
  const rbacPermissions = useRbacPermissions();
  console.log(rbacPermissions, "rbacPermissions");
  const authenticateUser = useAuthenticateUser();
  console.log(authenticateUser, "authenticateUser");

  return useQuery(
    'user',
    () =>
      Promise.all([authenticateUser, rbacPermissions]).then(
        ([userStatus, rbacPermissions]) => ({
          accountNumber: userStatus?.data.identity?.account_number,
          orgId: userStatus?.data.identity?.internal?.org_id,
          rbacPermissions: rbacPermissions?.data,
        })
      ),
    {
      enabled: rbacPermissions.isSuccess,
    }
  );
};

export default useUser;
