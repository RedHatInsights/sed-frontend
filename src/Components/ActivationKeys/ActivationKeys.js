import React, { useState } from 'react';
import {
  ActionGroup,
  Text,
  TextContent,
  TextVariants,
  PageSection,
  PageSectionVariants,
  Flex,
  FlexItem,
  Split,
  SplitItem,
} from '@patternfly/react-core';
import { Main } from '@redhat-cloud-services/frontend-components/Main';
import {
  PageHeader,
  PageHeaderTitle,
} from '@redhat-cloud-services/frontend-components/PageHeader';
import useChrome from '@redhat-cloud-services/frontend-components/useChrome';
import ActivationKeysTable from '../ActivationKeysTable';
import { useQueryClient } from 'react-query';
import NoActivationKeysFound from '../EmptyState';
import EditActivationKeyModal from '../Modals/EditActivationKeyModal';
import CreateActivationKeyWizard from '../Modals/CreateActivationKeyWizard';
import useActivationKeys from '../../hooks/useActivationKeys';
import Loading from '../LoadingState/Loading';
import CreateActivationKeyButton from './CreateActivationKeyButton';
import DeleteActivationKeyConfirmationModal from '../Modals/DeleteActivationKeyConfirmationModal';
import ActivationKeysDocsPopover from '../ActivationKeysDocsPopover';
const ActivationKeys = () => {
  const { updateDocumentTitle } = useChrome();
  updateDocumentTitle?.('Activation Keys - Remote Host Configuration');
  const queryClient = useQueryClient();
  const user = queryClient.getQueryData('user');
  const { isLoading, error, data } = useActivationKeys();
  const [isOpen, setisOpen] = useState(false);
  const [currentKeyName, setCurrentKeyName] = useState('');

  const [isDeleteActivationKeyModalOpen, setIsDeleteActivationKeyModalOpen] =
    useState(false);
  const [isEditActivationKeyModalOpen, setIsEditActivationKeyModalOpen] =
    useState(false);
  const handleModalToggle = () => {
    setisOpen(!isOpen);
  };

  const popoverContent = (
    <TextContent className="pf-v5-u-font-size-sm">
      <Text>
        Activation keys assist you in registering systems. Metadata such as
        role, system purpose, and usage can be automatically attached to systems
        via an activation key, and monitored with &nbsp;
        <a
          target="_blank"
          rel="noopener noreferrer"
          href={'https://console.redhat.com/insights/subscriptions/rhel'}
        >
          Subscription Watch.
        </a>
      </Text>
      <Text>
        To register with an activation key, you will need your organization ID:{' '}
        <b>{user.orgId}</b>
      </Text>
    </TextContent>
  );

  const setKeyName = (modalOpen, name) => {
    let currentName = modalOpen ? '' : name;
    setCurrentKeyName(currentName);
  };

  const handleDeleteActivationKeyModalToggle = (name) => {
    setKeyName(isDeleteActivationKeyModalOpen, name);
    setIsDeleteActivationKeyModalOpen(!isDeleteActivationKeyModalOpen);
  };

  const handleEditActivationKeyModalToggle = (name) => {
    setKeyName(isEditActivationKeyModalOpen, name);
    setIsEditActivationKeyModalOpen(!isEditActivationKeyModalOpen);
  };

  return (
    <React.Fragment>
      <PageHeader>
        <Split hasGutter className="page-title">
          <SplitItem isFilled>
            <Flex>
              <FlexItem spacer={{ default: 'spacerSm' }}>
                <PageHeaderTitle title="Activation Keys" />
              </FlexItem>
              <FlexItem>
                <ActivationKeysDocsPopover
                  popoverContent={popoverContent}
                  title="Activation Keys"
                  position="right"
                />
              </FlexItem>
            </Flex>
          </SplitItem>
        </Split>
        <TextContent>
          <Text component={TextVariants.p}>Organization ID: {user.orgId}</Text>
        </TextContent>
      </PageHeader>
      <Main>
        <PageSection variant={PageSectionVariants.light}>
          {isLoading && <Loading />}
          {!isLoading && !error && data.length > 0 && (
            <>
              <ActionGroup>
                <CreateActivationKeyButton onClick={handleModalToggle} />
              </ActionGroup>
              <ActivationKeysTable
                onDelete={handleDeleteActivationKeyModalToggle}
              />
            </>
          )}
          {!isLoading && !error && !data.length && (
            <NoActivationKeysFound handleModalToggle={handleModalToggle} />
          )}
        </PageSection>
      </Main>
      <CreateActivationKeyWizard
        key={isOpen}
        isOpen={isOpen}
        handleModalToggle={handleModalToggle}
      />
      <EditActivationKeyModal
        title="Edit activation key"
        isOpen={isEditActivationKeyModalOpen}
        handleModalToggle={handleEditActivationKeyModalToggle}
        activationKeyName={currentKeyName}
        modalSize="large"
      />
      <DeleteActivationKeyConfirmationModal
        handleModalToggle={handleDeleteActivationKeyModalToggle}
        isOpen={isDeleteActivationKeyModalOpen}
        name={currentKeyName}
      />
    </React.Fragment>
  );
};

export default ActivationKeys;
