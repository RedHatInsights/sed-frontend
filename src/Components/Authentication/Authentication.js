import React, { useEffect } from 'react';
import { useQueryClient } from '@tanstack/react-query';
import { useLocation } from 'react-router-dom';
import useUser from '../../hooks/useUser';
import Loading from '../LoadingState/Loading';
import Unavailable from '@redhat-cloud-services/frontend-components/Unavailable';
import propTypes from 'prop-types';
import useChrome from '@redhat-cloud-services/frontend-components/useChrome';
import NotAuthorized from '@redhat-cloud-services/frontend-components/NotAuthorized';

const Authentication = ({ children }) => {
  const queryClient = useQueryClient();
  const location = useLocation();
  const chrome = useChrome();
  const { isLoading, isFetching, isSuccess, isError, data } = useUser();
  const hasAnyPermission =
    data?.rbacPermissions &&
    (data.rbacPermissions.canReadActivationKeys ||
      data.rbacPermissions.canWriteActivationKeys);

  useEffect(() => {
    isSuccess && chrome?.hideGlobalFilter();
  }, [isSuccess]);

  useEffect(() => {
    /**
     * On every rerender, based on URL change (location.pathname),
     * reset the user's status to loading before authenticating again.
     */
    queryClient.invalidateQueries(['user']);
  }, [location.pathname]);

  if (isError === true) {
    return <Unavailable />;
  } else if (isLoading === true || isFetching === true) {
    return <Loading />;
  } else if (isSuccess === true && !hasAnyPermission) {
    return <NotAuthorized serviceName="Remote Host Configuration" />;
  } else if (isSuccess === true) {
    return <>{children}</>;
  } else {
    return <Loading />;
  }
};

Authentication.propTypes = {
  children: propTypes.node,
};

export default Authentication;
