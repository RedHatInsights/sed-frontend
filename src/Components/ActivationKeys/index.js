import ActivationKeys from './ActivationKeys';

export default ActivationKeys;
