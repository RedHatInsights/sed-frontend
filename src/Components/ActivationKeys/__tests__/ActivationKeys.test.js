import React from 'react';
import { render, waitFor, screen } from '@testing-library/react';
import ActivationKeys from '../index';
import { Provider } from 'react-redux';
import Authentication from '../../../Components/Authentication';
import { BrowserRouter as Router } from 'react-router-dom';
import { init } from '../../../store';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import useUser from '../../../hooks/useUser';
import { get, def } from 'bdd-lazy-var';
import useActivationKeys from '../../../hooks/useActivationKeys';
import '@testing-library/jest-dom';
jest.mock('../../../hooks/useActivationKeys');
jest.mock('../../../hooks/useUser');
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => ({
    pathname: '/',
  }),
}));

const queryClient = new QueryClient();

const PageContainer = () => (
  <QueryClientProvider client={queryClient}>
    <Authentication>
      <Provider store={init().getStore()}>
        <Router>
          <ActivationKeys />
        </Router>
      </Provider>
    </Authentication>
  </QueryClientProvider>
);

const mockAuthenticateUser = (isLoading, isError, rbacPermissions) => {
  const user = {
    accountNumber: '123',
    orgId: '123',
    rbacPermissions: rbacPermissions,
  };
  useUser.mockReturnValue({
    isLoading: isLoading,
    isFetching: false,
    isSuccess: true,
    isError: isError,
    data: user,
  });

  if (isError === false) {
    queryClient.setQueryData(['user'], user);
  }
};

// eslint-disable-next-line react/display-name
jest.mock('../../../Components/ActivationKeysTable', () => () => (
  <div>Activation Keys Table</div>
));
jest.mock(
  '@redhat-cloud-services/frontend-components/NotAuthorized',
  // eslint-disable-next-line react/display-name
  () => () => <div>Not Authorized</div>
);

jest.mock(
  '@redhat-cloud-services/frontend-components/Unavailable',
  // eslint-disable-next-line react/display-name
  () => () => <div>Unavailable</div>
);

describe('ActivationKeys', () => {
  def('isLoading', () => false);
  def('isError', () => false);
  def('rbacPermissions', () => {
    return { canReadActivationKeys: true, canWriteActivationKeys: true };
  });
  def('keysData', () => [
    {
      name: 'A',
      role: 'B',
      serviceLevel: 'C',
      usage: 'D',
    },
  ]);

  beforeEach(() => {
    jest.resetAllMocks();
    mockAuthenticateUser(
      get('isLoading'),
      get('isError'),
      get('rbacPermissions')
    );
    useActivationKeys.mockReturnValue({
      isLoading: false,
      isFetching: false,
      isError: false,
      isSuccess: true,
      data: get('keysData'),
    });
  });

  it('renders correctly', async () => {
    const { container } = render(<PageContainer />);
    await waitFor(() => expect(useUser).toHaveBeenCalledTimes(1));
    expect(container).toMatchSnapshot();
  });

  describe('when the user call fails', () => {
    def('isError', () => true);

    it('renders an error message when user call fails', async () => {
      const { container } = render(<PageContainer />);
      await waitFor(() => expect(useUser).toHaveBeenCalledTimes(1));
      expect(container).toMatchSnapshot();
    });
  });

  describe('when the user does not have proper permissions', () => {
    def('rbacPermissions', () => {
      return { canReadActivationKeys: false };
    });

    it('redirects to not authorized page', async () => {
      const { container } = render(<PageContainer />);
      await waitFor(() => expect(useUser).toHaveBeenCalledTimes(1));
      expect(container).toMatchSnapshot();
    });
  });

  describe('when the user have only read permissions', () => {
    def('rbacPermissions', () => {
      return { canReadActivationKeys: true, canWriteActivationKeys: false };
    });

    it('create activation key button is disabled', async () => {
      render(<PageContainer />);
      await waitFor(() => expect(useUser).toHaveBeenCalledTimes(1));
      expect(screen.getByText('Create activation key')).toBeDisabled();
    });
  });

  describe('show blank state when no activation keys', () => {
    def('keysData', () => []);

    it('renders blank state', async () => {
      render(<PageContainer />);
      expect(screen.getByText('No activation keys')).toBeInTheDocument();
    });
  });
});
