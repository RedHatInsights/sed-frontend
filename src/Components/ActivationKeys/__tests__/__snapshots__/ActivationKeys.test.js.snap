// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivationKeys renders correctly 1`] = `
<div>
  <section
    class="pf-v5-l-page-header pf-v5-c-page-header pf-v5-l-page__main-section pf-v5-c-page__main-section pf-m-light"
    widget-type="InsightsPageHeader"
  >
    <div
      class="pf-v5-l-split pf-m-gutter page-title"
    >
      <div
        class="pf-v5-l-split__item pf-m-fill"
      >
        <div
          class="pf-v5-l-flex"
        >
          <div
            class="pf-m-spacer-sm"
          >
            <h1
              class="pf-v5-c-title pf-m-2xl"
              data-ouia-component-id="OUIA-Generated-Title-1"
              data-ouia-component-type="PF5/Title"
              data-ouia-safe="true"
              widget-type="InsightsPageHeaderTitle"
            >
              Activation Keys
            </h1>
          </div>
          <div
            class=""
          >
            <div
              style="display: contents;"
            >
              <button
                aria-disabled="false"
                class="pf-v5-c-button pf-m-plain"
                data-ouia-component-id="OUIA-Generated-Button-plain-1"
                data-ouia-component-type="PF5/Button"
                data-ouia-safe="true"
                style="padding: 0px;"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 512 512"
                  width="1em"
                >
                  <path
                    d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="pf-v5-c-content"
    >
      <p
        class=""
        data-ouia-component-id="OUIA-Generated-Text-1"
        data-ouia-component-type="PF5/Text"
        data-ouia-safe="true"
        data-pf-content="true"
      >
        Organization ID: 
        123
      </p>
    </div>
  </section>
  <section
    class="pf-v5-l-page__main-section pf-v5-c-page__main-section"
  >
    <section
      class="pf-v5-c-page__main-section pf-m-light"
    >
      <div
        class="pf-v5-c-form__group pf-m-action"
      >
        <div
          class="pf-v5-c-form__group-control"
        >
          <div
            class="pf-v5-c-form__actions"
          >
            <button
              aria-disabled="false"
              class="pf-v5-c-button pf-m-primary"
              data-ouia-component-id="OUIA-Generated-Button-primary-1"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe="true"
              type="button"
            >
              Create activation key
            </button>
          </div>
        </div>
      </div>
      <div>
        Activation Keys Table
      </div>
    </section>
  </section>
</div>
`;

exports[`ActivationKeys when the user call fails renders an error message when user call fails 1`] = `
<div>
  <div>
    Unavailable
  </div>
</div>
`;

exports[`ActivationKeys when the user does not have proper permissions redirects to not authorized page 1`] = `
<div>
  <div>
    Not Authorized
  </div>
</div>
`;
